src/Sylius/Component/Core/Dashboard/SalesDataProvider.php

```diff
@@ -13,51 +13,60 @@
 
 /**
  * @experimental
  */
 final class SalesDataProvider implements SalesDataProviderInterface
 {
-    /** @var EntityManagerInterface */
-    private $entityManager;
+    /** @var OrderRepositoryInterface|EntityRepository */
+    private $orderRepository;
 
-    public function __construct(EntityManagerInterface $entityManager)
+    public function __construct(OrderRepositoryInterface $orderRepository)
     {
-        $this->entityManager = $entityManager;
+        $this->orderRepository = $orderRepository;
     }
 
-    public function getLastYearSalesSummary(ChannelInterface $channel): SalesSummaryInterface
-    {
-        $startDate = (new \DateTime('first day of next month last year'))->format('Y/m/d');
-        $endDate = (new \DateTime('last day of this month'))->format('Y/m/d');
-        $channelId = $channel->getId();
-
-        $query = $this->entityManager->getConnection()->query(
-            "SELECT
-                DATE_FORMAT(checkout_completed_at, '%m.%y') AS \"date\",
-                SUM(total) as \"total\"
-            FROM sylius_order
-            WHERE (channel_id = $channelId)
-            AND (checkout_completed_at BETWEEN '$startDate' AND '$endDate')
-            AND (payment_state = 'paid')
-            GROUP BY date;"
-        );
-
-        $query->execute();
-        $result = $query->fetchAll();
+    function getSalesSummary(
+        \DateTimeInterface $startDate,
+        \DateTimeInterface $endDate,
+        string $interval,
+        ChannelInterface $channel,
+        string $dateFormat
+    ): SalesSummaryInterface {
+        $ordersTotals = $this->orderRepository->createQueryBuilder('o')
+            ->select('HOUR(o.checkoutCompletedAt) AS hour')
+            ->addSelect('DAY(o.checkoutCompletedAt) AS day')
+            ->addSelect('MONTH(o.checkoutCompletedAt) AS month')
+            ->addSelect('YEAR(o.checkoutCompletedAt) AS year')
+            ->addSelect('SUM(o.total) AS total')
+            ->where('o.checkoutCompletedAt >= :startDate')
+            ->andWhere('o.checkoutCompletedAt <= :endDate')
+            ->andWhere('o.paymentState = :state')
+            ->andWhere('o.channel = :channel')
+            ->setParameter('startDate', $startDate)
+            ->setParameter('endDate', $endDate)
+            ->setParameter('state', OrderPaymentStates::STATE_PAID)
+            ->setParameter('channel', $channel)
+            ->groupBy($interval)
+            ->addGroupBy('o.checkoutCompletedAt')
+            ->getQuery()
+            ->getResult()
+        ;
 
         $data = [];
-        foreach ($result as $item) {
-            $data[$item['date']] = (int) $item['total'];
+
+        foreach ($ordersTotals as $item) {
+            $data[(string) str_pad($item[$interval],2,"0",STR_PAD_LEFT)] = 0;
+        }
+        foreach ($ordersTotals as $item) {
+            $data[(string) str_pad($item[$interval],2,"0",STR_PAD_LEFT)] += (int) $item['total'];
         }
```

Comment:
This should provide the same end result: 
```suggestion
        foreach ($ordersTotals as $item) {
            $data[(string) str_pad($item[$interval],2,"0",STR_PAD_LEFT)] = (int) $item['total'];
        }
```

Reply:
i am afraid it is not the best suggestion :(
because your solution overriding the previous value, my solution is adding the next value to the previous one,
but you are right it should be refactored :)

Reply:
now I see it. `array_reduce` could do something like this but we need to filter data by intervals first anyway